### PR TITLE
Add Enterprise SONiC "dhcp_relay" resource module

### PR DESCRIPTION
**Pull request by ArunSaravananBalachandran**
##### SUMMARY

- Add Enterprise SONiC 'dhcp_relay' resource module:
  - Provide data definitions in 'argspec'.
  - Implement 'facts' to gather facts for the module.
  - Implement 'config' to apply configurations for `merged` and `deleted` states.
- Define regression test cases to verify correct functionality of dhcp_relay resource module.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

- sonic_dhcp_relay

#### Test Results

- Regression test HTML report:
[dhcp_relay_regression_report.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/9936273/dhcp_relay_regression_report.pdf)

- Regression test detailed log:
[dhcp_relay_regression_terminal_output.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/9936270/dhcp_relay_regression_terminal_output.txt)

- Show running config and respective facts gathered:
[show_running_get_facts_output.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/9936282/show_running_get_facts_output.txt)

### CODE REVIEW

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+        ipv4 = command.get('ipv4')
+        have_ipv4 = config.get('ipv4')
+        if not ipv4 or not have_ipv4:
+            return requests
+
+        server_addresses = self.get_server_addresses(ipv4.get('server_addresses'))
+        have_server_addresses = self.get_server_addresses(have_ipv4.get('server_addresses'))
+
+        # Delete all DHCP relay config for an interface, if only
+        # an address with no value is specified.
```
Comment (kerry-meyer) [inline plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py:379]: Please add an additional explanatory sentence here and for the corresponding DHCPv6 block:

e.g.

This "special" YAML sequence is supported to provide "delete all AF parameters" functionality despite the Ansible infrastructure limitations that prevent use of a simpler syntax for deleting an entire AF parameter dictionary.

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+            for addr in del_server_addresses:
+                url = self.dhcp_relay_intf_config_path['server_address'].format(intf_name=name, server_address=addr)
+                requests.append({'path': url, 'method': DELETE})
+
+        # Specifying appropriate order for deletion to succeed
+        if ipv4.get('link_select') is not None and have_ipv4.get('link_select'):
+            url = self.dhcp_relay_intf_config_path['link_select'].format(intf_name=name)
+            requests.append({'path': url, 'method': DELETE})
+
+        if ipv4.get('source_interface') and have_ipv4.get('source_interface'):
```
Comment (kerry-meyer) [inline plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py:402]: Please add verification that the source addresses match before enqueuing this deletion request.

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+            url = self.dhcp_relay_intf_config_path['source_interface'].format(intf_name=name)
+            requests.append({'path': url, 'method': DELETE})
+
+        if (ipv4.get('max_hop_count') and have_ipv4.get('max_hop_count')
+                and have_ipv4.get('max_hop_count') != DEFAULT_MAX_HOP_COUNT):
```
Comment (kerry-meyer) [inline plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py:407]: For deletion, the specified value is required to match the configured value. Please add verification before enqueuing the deletion request.

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+            url = self.dhcp_relay_intf_config_path['max_hop_count'].format(intf_name=name)
+            requests.append({'path': url, 'method': DELETE})
+
+        if ipv4.get('vrf_select') is not None and have_ipv4.get('vrf_select'):
+            url = self.dhcp_relay_intf_config_path['vrf_select'].format(intf_name=name)
+            requests.append({'path': url, 'method': DELETE})
+
+        if (ipv4.get('policy_action') and have_ipv4.get('policy_action')
+                and have_ipv4.get('policy_action') != DEFAULT_POLICY_ACTION):
```
Comment (kerry-meyer) [inline plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py:416]: Please verify matching "have" and "want" configuration before enqueuing the request.

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+            url = self.dhcp_relay_intf_config_path['policy_action'].format(intf_name=name)
+            requests.append({'path': url, 'method': DELETE})
+
+        if (ipv4.get('circuit_id') and have_ipv4.get('circuit_id')
+                and have_ipv4.get('circuit_id') != DEFAULT_CIRCUIT_ID):
```
Comment (kerry-meyer) [inline plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py:421]: Please verify matching "have" and "want" configuration before enqueuing the request.

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+            url = self.dhcp_relay_intf_config_path['circuit_id'].format(intf_name=name)
+            requests.append({'path': url, 'method': DELETE})
+
+        return requests
+
+    def get_delete_specific_dhcpv6_relay_param_requests(self, command, have):
```
Comment (kerry-meyer) [inline plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py:427]: The same comments posted for "get_delete_specific_dhcp_relay_param_request" apply to this function.

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+        have_server_addresses = self.get_server_addresses(have_ipv4.get('server_addresses'))
+
+        # Delete all DHCP relay config for an interface, if only
+        # an address with no value is specified.
+        if (ipv4.get('server_addresses') and len(ipv4.get('server_addresses'))
+                and not server_addresses):
+            requests.append({'path': self.dhcp_relay_intf_config_path['server_addresses_all'].format(intf_name=name), 'method': DELETE})
+            return requests
+
+        del_server_addresses = have_server_addresses.intersection(server_addresses)
```
Comment (kerry-meyer) [inline plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py:385]: This is a good way of making use of Python's "set" functionality.

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+        if del_server_addresses:
+            # Delete all DHCP relay config for an interface if all
+            # server addresses are to be deleted
```
Comment (kerry-meyer) [inline plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py:388]: Please add a sentence to explain the rationale for deleting all parameters in this case. (It's fine to do it, but should be explained in the comment.)

File: tests/regression/roles/sonic_dhcp_relay/defaults/main.yml [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+        ipv6:
+          server_addresses:
+            - address: 101::3
+          max_hop_count: 8
+      - name: '{{ vlan1 }}'
+        ipv4:
+          policy_action: 'append'
+
+  - name: test_case_06
+    description: Delete specific DHCP and DHCPv6 relay configurations
```
Comment (kerry-meyer) [inline tests/regression/roles/sonic_dhcp_relay/defaults/main.yml:192]: Please describe this test case more specifically. (I assume it is intended to verify that deletion of all server addresses results in deletion of the other parameters for the affected interface/AF as well.)

**Review by kerry-meyer — CHANGES_REQUESTED** [review]
This is a great job on your first resource module! The code structure looks good everywhere. My current comments are either minor or related to documentation clarification.

Although I have not finished reading carefully through all of the regression test related files, I have done a quick scan of them and am not expecting many, if any, additional comments.

I am posting a preliminary set of comments for the files for which I have completed first-pass reviewing to enable you to address or reply to the first set of comments while I check the remaining files.

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+            have_obj = next((cfg for cfg in have if cfg['name'] == intf_name), None)
+            if not have_obj:
+                continue
+
+            have_ipv4 = have_obj.get('ipv4')
+            have_ipv6 = have_obj.get('ipv6')
+
+            ipv4 = command.get('ipv4')
+            ipv6 = command.get('ipv6')
+            if intf_name and not ipv4 and not ipv6:
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py]: As suggested, have removed intf_name.

File: plugins/modules/sonic_dhcp_relay.py [hunk 11]
```diff
@@ -0,0 +1,554 @@
...
+#  speed 400000
+#  fec RS
+#  no shutdown
+#  ip address 61.1.1.1/24
+#  ip dhcp-relay 71.1.1.1 72.1.1.1 73.1.1.1
+#  ip dhcp-relay source-interface Vlan100
+#  ip dhcp-relay link-select
+#  ip dhcp-relay circuit-id %h:%p
+# !
+
```
[the remark this answers is not in the document]
Comment (ArunSaravananBalachandran) [reply plugins/modules/sonic_dhcp_relay.py:272]: Have updated the example.

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py (hunk 1 above)
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py]: Comment has been updated

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py (hunk 2 above)
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py]: Done

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py (hunk 3 above)
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py]: Done

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py (hunk 4 above)
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py]: Done

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py (hunk 5 above)
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py]: Done

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py (hunk 6 above)
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py:438]: Condition to check for matching have and want attributes has been added.

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py (hunk 8 above)
Comment (ArunSaravananBalachandran) [reply plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py]: Updated the comment based on the latest request set for deletion.

File: tests/regression/roles/sonic_dhcp_relay/defaults/main.yml (hunk 9 above)
Comment (ArunSaravananBalachandran) [reply tests/regression/roles/sonic_dhcp_relay/defaults/main.yml]: Have updated the description for the test cases 6, 7, and 8 achieving similar functionality with different approaches

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py]: Thanks.

Resolved.

File: plugins/modules/sonic_dhcp_relay.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (kerry-meyer) [reply plugins/modules/sonic_dhcp_relay.py:272]: Thanks.

Resolved.

File: plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py (hunk 1 above)
Comment (kerry-meyer) [reply plugins/module_utils/network/sonic/config/dhcp_relay/dhcp_relay.py]: Thanks.

Resolved.

File: tests/regression/roles/sonic_dhcp_relay/defaults/main.yml (hunk 9 above)
Comment (kerry-meyer) [reply tests/regression/roles/sonic_dhcp_relay/defaults/main.yml]: The updates look good.

Thanks.

**Review by kerry-meyer — APPROVED** [review]
All requested review changes have been made and the updated code looks good.

I have finished reviewing the remaining test "task" files and they all look good as well.

Great job on this resource module!

Approved.

**Pull request merged**